### PR TITLE
Add possibility to zero out memory before returning it to the pool

### LINKED ISSUE #267: FeatureRequest: Add opt-in possibility to zeroed-out buffers
Hi together,

Would it be possible to extend the RecyclableMemoryStreamManager by a configuration option to zero-out buffers before returning them to the pool?
I understand that for performance reasons, the buffers are not ever pre-initialized or zeroed-out by default, but i think it would be nice to have the option to configure it if one is willing to take the performance hit.

I could create a PR for this if you think it would be a useful addition.

Best regards,
Lars

### PR DESCRIPTION
**Pull request by lklein53**
Closes #267 
TODO

- ~~Add test cases~~

### CODE REVIEW

**Comment by benmwatson** [comment]
@lklein53 Are you still working on this PR? I've left a comment, and it needs test cases.

**Comment by lklein53** [comment]
> @lklein53 Are you still working on this PR? I've left a comment, and it needs test cases.

Sorry completly forgot about this. Will pick it up in 2 weeks once i am back from vacation. However i can't see the comment you mentioned.

File: src/RecyclableMemoryStreamManager.cs [hunk 1]
```diff
@@ -368,6 +368,12 @@ public long LargeBuffersFree
         /// set and <see cref="RecyclableMemoryStream.ToArray"/> is called, a <c>NotSupportedException</c> will be thrown.</remarks>
         public bool ThrowExceptionOnToArray { get; set; }
 
+        /// <summary>
+        /// Zero out buffers before returning them to the pool.
+        /// </summary>
+        /// <remarks>Setting this to true causes a performance hit and should only be set if one wants to avoid accidental data leaks.</remarks>
```
Comment (grbell-ms) [inline src/RecyclableMemoryStreamManager.cs:374]: This flag does more than the comment suggests. It zeros newly allocated buffers, as well as returned buffers.

File: src/RecyclableMemoryStreamManager.cs [hunk 2]
```diff
@@ -632,6 +639,12 @@ internal void ReturnBlock(byte[] block, Guid id, string tag)
             }
         }
 
+        private void ZeroOutMemoryIfEnabled(Span<byte> buffer) {
+            if (this.ZeroOutBuffer) {
+                buffer.Clear();
+            }
+        }
```
Comment (grbell-ms) [inline src/RecyclableMemoryStreamManager.cs:646]: We don't want to construct a span when we don't have to.

```suggestion
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private void ZeroOutMemoryIfEnabled(byte[] buffer) {
            if (this.ZeroOutBuffer) {
                Array.Clear(buffer);
            }
        }
```

Comment (lklein53) [reply src/RecyclableMemoryStreamManager.cs]: Thanks. Chankged the code

File: src/RecyclableMemoryStreamManager.cs (hunk 1 above)
Comment (lklein53) [reply src/RecyclableMemoryStreamManager.cs:374]: Thanks. Adjusted the comment

Comment (lklein53) [reply src/RecyclableMemoryStreamManager.cs:374]: Thanks. Adjusted the comment

File: src/RecyclableMemoryStreamManager.cs [hunk 3]
```diff
@@ -460,9 +466,9 @@ internal byte[] GetLargeBuffer(long requiredSize, Guid id, string tag)
             return buffer;
 
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            static byte[] AllocateArray(long requiredSize) =>
+            static byte[] AllocateArray(long requiredSize, bool zeroInitializeArray) =>
 #if NET5_0_OR_GREATER
-                GC.AllocateUninitializedArray<byte>((int)requiredSize);
+                zeroInitializeArray ? GC.AllocateArray<byte>((int)requiredSize) : GC.AllocateUninitializedArray<byte>((int)requiredSize);
 #else
                 new byte[requiredSize];
```
Comment (benmwatson) [inline src/RecyclableMemoryStreamManager.cs:473]: What about the < .NET 5.0 case?

**Review by benmwatson — CHANGES_REQUESTED** [review]

**Review by benmwatson — APPROVED** [review]

**Comment by lklein53** [comment]
It seems like the pipeline failed with the following message

> [error]The runner has received a shutdown signal. This can happen when the runner service is stopped, or a manually started runner is canceled.

I haven't found a way to manually trigger the CI Pipeline again. Could you tell me how to do this?

**Comment by benmwatson** [comment]
> I haven't found a way to manually trigger the CI Pipeline again. Could you tell me how to do this?

I'm able to re-run the pipeline, but it hasn't helped. I thought there could be some broken unit tests, but they all run fine on my local machine form your repo.

I wonder if it's a problem with some tests that allocate a ton of memory, and maybe something in the new functionality is causing the test process to crash or exceed some resource constraints.

I can't figure out how to get the logs it mentions. I'll keep digging at it.

**Comment by lklein53** [comment]
> > I haven't found a way to manually trigger the CI Pipeline again. Could you tell me how to do this?
> 
> I'm able to re-run the pipeline, but it hasn't helped. I thought there could be some broken unit tests, but they all run fine on my local machine form your repo.
> 
> I wonder if it's a problem with some tests that allocate a ton of memory, and maybe something in the new functionality is causing the test process to crash or exceed some resource constraints.
> 
> I can't figure out how to get the logs it mentions. I'll keep digging at it.

Thanks a lot. Yeah i also tried them locally and they run fine.

**Comment by benmwatson** [comment]
I've done some digging on this and have discovered the following:

* The test suite will pass if I disable your new test configuration.
* The test suite will pass if I disable all the VeryLargeStream_* tests. 
* Monitoring the test suites on my dev machines shows that the new test configuration uses more memory (~26GB) than the others (~20GB).
* I'm not sure if I can get access to larger runners. I'll have to figure out who to ask on the Microsoft account about this.
* I'm assuming there's something about zeroing out the memory that is causing more memory to actually be committed. In many of the tests, I don't think we're actually explicitly touching all the buffers, so this could be very possible.

My immediate recommendation is that you add a switch to the tests that will skip running the VeryLargeStream* tests in your test configuration.

**Comment by lklein53** [comment]
Thanks for diggin into it. I went with your suggestion

**Review by benmwatson — APPROVED** [review]

**Pull request merged**